### PR TITLE
Momentum Optimizer: Now supports Nesterov flag

### LINKED ISSUE #728: Add use_nesterov flag to MomentumOptimizer
Currently, `MomentumOptimizer` does not implement Nesterov momentum which [Tensorflow](https://www.tensorflow.org/api_docs/python/tf/train/MomentumOptimizer) supports. Need to add this feature.

### PR DESCRIPTION
**Pull request by ManrajGrover**
This PR adds support for Nesterov Accelerated Gradient to Momentum Optimizer

Reference: http://proceedings.mlr.press/v28/sutskever13.pdf

Fixes #728

---
This change is [Reviewable](https://reviewable.io/reviews/pair-code/deeplearnjs/764)

### CODE REVIEW

**Comment by ManrajGrover** [comment]
Adding screenshot here:

screen shot 2018-02-24 at 5 51 18 pm

**Comment by nsthorat** [comment]
Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[src/optimizers/momentum_optimizer.ts, line 59 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/764#-L67NSAJ3P_-JLLO7xt5:-L67NSAJ3P_-JLLO7xt6:b-7rk6em) ([raw file](https://github.com/pair-code/deeplearnjs/blob/9bbaf94bc48b544aa0e2de1e10520dedf2787c10/src/optimizers/momentum_optimizer.ts#L59)):*
> ```XML
> 
>       tidy(() => {
>         let newValue: Variable;
> ```

this should just be Tensor

---

*[src/optimizers/momentum_optimizer.ts, line 62 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/764#-L67PKu97LrlJ7gXNgdD:-L67PKu97LrlJ7gXNgdE:b6v5htn) ([raw file](https://github.com/pair-code/deeplearnjs/blob/9bbaf94bc48b544aa0e2de1e10520dedf2787c10/src/optimizers/momentum_optimizer.ts#L62)):*
> ```XML
>         const newAccumulation = this.m.mul(accumulation).add(gradient);
>         if (this.useNesterov) {
>           newValue =
> ```

I don't quite follow this math here, it looks like we need to compute gradients at the "look a head", instead of the current position, or you can make it look like vanilla SGD as karpathy says here:

http://cs231n.github.io/neural-networks-3/

```
v_prev = v # back this up
v = mu * v - learning_rate * dx # velocity update stays the same
x += -mu * v_prev + (1 + mu) * v # position update changes form
```

This looks similar, but not exactly the same. Am I misunderstanding?

---

*[src/optimizers/momentum_optimizer.ts, line 102 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/764#-L67NUqy4KOORuk3lrZy:-L67NUqy4KOORuk3lrZz:b-e1jc0r) ([raw file](https://github.com/pair-code/deeplearnjs/blob/9bbaf94bc48b544aa0e2de1e10520dedf2787c10/src/optimizers/momentum_optimizer.ts#L102)):*
> ```XML
>         const oldVelocity = this.variableVelocitiesGraph.get(node.output);
> 
>         let variable: Variable;
> ```

This should be a Tensor

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/764)*

**Comment by ManrajGrover** [comment]
@nsthorat I'm not sure. Doing the math below:

Equations shared in [this](http://cs231n.github.io/neural-networks-3/) article:

```
v_prev = v
v = mu * v - lr * dx
x += -mu * v_prev + (1 + mu) * v
=> x += -mu * v_prev + (1 + mu) * (mu * v_prev - lr * dx)
=> x += -mu * v_prev + mu * v_prev - lr * dx + mu^2 * v_prev - lr * dx * mu
=> x += - lr * dx + mu^2 * v_prev - lr * dx * mu
=> x +=  mu^2 * v_prev - (1 + mu) * lr * dx
```

Equations implemented:

```
temp = mu * v_prev + dx
=> x += -lr * (dx + temp * mu)
=> x += -lr * dx - lr * temp * mu
=> x += -lr * dx - lr * (mu * v_prev + dx) * mu
=> x += -lr * dx - mu^2 * lr * v_prev - lr * mu * dx
=> x += mu^2 * v_prev * lr - (1 + mu) * lr * dx
```
The only difference I see is `lr` multiplied in first term. I had actually double checked with Tensorflow [implementation](https://github.com/tensorflow/tensorflow/blob/982549ea3423df4270ff154e5c764beb43d472da/tensorflow/core/kernels/training_ops_gpu.cu.cc#L93) and fixed this. Could you please check the same as well?

**Comment by nsthorat** [comment]
:lgtm_strong:

Sorry for the delay. LGTM, I also looked at TF implementation and this looks identical.

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/764#-:-L6mw-tABHCoBDGIEMyL:b-k1a38o)*
